src/rust/engine/graph/src/lib.rs

```diff
@@ -177,40 +306,94 @@ impl<N: Node> Entry<N> {
     match &self.state {
-      &EntryState::Completed { ref result, .. } => Some(result.clone()),
+      &EntryState::Completed {
+        ref result, dirty, ..
+      } if !dirty =>
+      {
+        Some(result.clone())
+      }
       _ => None,
     }
   }
 
   ///
   /// Called from the Executor when a Node completes.
   ///
-  fn complete(&mut self, result_run_token: RunToken, result: Result<N::Item, N::Error>) {
-    // Temporarily swap in `NotStarted` in order to take ownership of the state.
-    let state = mem::replace(&mut self.state, EntryState::initial());
-
+  /// A `result` value of `None` indicates that the Node was found to be clean, and its previous
+  /// result should be used. This special case exists to avoid 1) cloning the result to call this
+  /// method, and 2) comparing the current/previous results unnecessarily.
+  ///
+  fn complete<C>(
+    &mut self,
+    context: &C,
+    entry_id: EntryId,
+    result_run_token: RunToken,
+    result: Option<Result<N::Item, N::Error>>,
+  ) where
+    C: NodeContext<Node = N>,
+  {
     // We care about exactly one case: a Running state with the same run_token. All other states
     // represent various (legal) race conditions. See `RunToken`'s docs for more information.
-    self.state = match state {
+    match &self.state {
+      &EntryState::Running { run_token, .. } if result_run_token == run_token => {}
+      _ => {
+        // We care about exactly one case: a Running state with the same run_token. All other states
+        // represent various (legal) race conditions.
+        return;
+      }
+    }
+
+    self.state = match mem::replace(&mut self.state, EntryState::initial()) {
       EntryState::Running {
         waiters,
-        start_time,
         run_token,
+        generation,
+        previous_result,
+        dirty,
+        ..
       } => {
-        if result_run_token == run_token {
+        if dirty {
+          // The node was dirtied while we were running. The newly computed result value was
+          // never published, so we continue to use the previous result.
+          Self::start(
+            context,
+            &self.node,
+            entry_id,
+            run_token,
+            generation,
+            previous_result,
+          )
+        } else {
+          // If the new result does not match the previous result, the generation increments.
+          let (generation, next_result) = if let Some(result) = result {
+            if Some(&result) == previous_result.as_ref() {
+              // Node was re-executed, but had the same result value.
+              (generation, result)
+            } else {
+              (generation.next(), result)
+            }
+          } else {
+            // Node was marked clean.
+            // NB: The unwrap here avoids a clone and a comparison: see the method docs.
+            (
+              generation,
+              previous_result.unwrap_or_else(|| {
```

Comment:
`previous_result.expect("A node cannot be marked clean without a previous result.")`?